lkmpg.tex

```diff
@@ -1204,6 +1204,75 @@ \section{System Calls}
 So, if we want to change the way a certain system call works, what we need to do is to write our own function to implement it (usually by adding a bit of our own code, and then calling the original function) and then change the pointer at \cpp|sys_call_table| to point to our function.
 Because we might be removed later and we don't want to leave the system in an unstable state, it's important for \cpp|cleanup_module| to restore the table to its original state.
 
+To change the pointer in \cpp|sys_call_table|, we need to consider the control register.
+A control register is a processor register that changes or controls the general behavior of the CPU.
+For x86 architecture, the \verb|cr0| register has various control flags that modify the basic operation of the processor.
+The \verb|WP| flag in \verb|cr0| stands for write protection.
+Once the \verb|WP| flag is set, the CPU may not write into the read-only sections.
+To modify \cpp|sys_call_table|, we need to disable the \verb|WP| flag.
+Since Linux v5.3, the \cpp|write_cr0| function cannot be used because of the sensitive \verb|cr0| bits pinned by the security issue, the attacker may write into CPU control registers to disable CPU protections like read-only protection.
+So we have to provide the custom assembly routine to bypass it.
+
+However, \cpp|sys_call_table| symbol was unexported to prevent misuse.
+But there have few ways to get the symbol, manual symbol lookup, and \cpp|kallsyms_lookup_name|, here we use \cpp|kallsyms_lookup_name|.
+Unfortunately, since Linux v5.7, \cpp|kallsyms_lookup_name| is also unexported, it needs certain trick to get the address of \cpp|kallsyms_lookup_name|.
+If \cpp|CONFIG_KPROBES| is enabled, we can facilitate the retrieval of function addresses by means of Kprobes to dynamically break into the specific kernel routine.
+Kprobes inserts a breakpoint at the entry of function by replacing the first bytes of the probed instruction.
+When a CPU hits the breakpoint, registers are stored, and the control will pass to Kprobes.
+It passes the addresses of the saved registers and the Kprobe struct to the handler you defined, then executes it.
+Kprobes can be registered by symbol name or address.
+Within the symbol name, the address will be handled by the kernel.
+
+Otherwise, use manual symbol lookup or getting the address of \cpp|sys_call_table| from \verb|/proc/kallsyms| and \verb|/boot/System.map|.
+Following is the sample usage for finding the address from \verb|/proc/kallsyms|:
+\begin{verbatim}
+$ sudo grep sys_call_table /proc/kallsyms
+ffffffff82000280 R x32_sys_call_table
+ffffffff820013a0 R sys_call_table
+ffffffff820023e0 R ia32_sys_call_table
+$ insmod syscall.ko addr=0xffffffff820013a0
+\end{verbatim}
+Using the address from \verb|/boot/System.map|, be careful about the KASLR (Kernel Address Space Layout Randomization).
```

Comment:
Improve the writing. You can say,
>  if the address of a symbol in `/proc/kallsyms` is different from the address in System.map, KASLR is enabled with the kernel which your system running on.

See https://dev.to/satorutakeuchi/a-brief-description-of-aslr-and-kaslr-2bbp